### ISSUE TITLE
GroupBy with to-many relation leads to exception and overly complex SQL (regression since 6.0.2)

### ISSUE THREAD

**Issue by sgaertner** [dotnet/efcore#28309]
With 6.0.2 the following code leads to an exception with Npgsql and _at least_ an overly complex SQL with other providers. All is fine with previous EF core versions (3.1, 5.0, 6.0.1). The issue is probably caused by the fix for #27102.

### How to reproduce

Very simple example, using the EF core tutorial model: [GroupByExists.zip](https://github.com/dotnet/efcore/files/8974370/GroupByExists.zip)

The query in question:

```C#
// calculate count of blogs with and without any posts
var counts = context.Blogs
    .GroupBy(b => b.Posts.Any(), (k, ts) => new { Key = k, Count = ts.Count() })
    .ToDictionary(t => t.Key, t => t.Count);
```

This has worked flawlessly with Npgsql until EF core 6.0.1 and throws an exception with EF core 6.0.2:

```
fail: Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'GroupByExists.BloggingContext'.
      Npgsql.PostgresException (0x80004005): 42803: subquery uses ungrouped column "b.BlogId" from outer query
```

The exception is caused by an overly complex SQL, which is also generated with other providers (e.g. Sqlite).

### Expected SQL (as generated until EF core 6.0.1)

```SQL
SELECT
    EXISTS (SELECT 1 FROM "Posts" AS "p" WHERE "b"."BlogId" = "p"."BlogId") AS "Key",
    COUNT(*) AS "Count"
FROM "Blogs" AS "b"
GROUP BY
    EXISTS (SELECT 1 FROM "Posts" AS "p" WHERE "b"."BlogId" = "p"."BlogId")
```

### Generated SQL (since EF core 6.0.2)

```SQL
SELECT
    EXISTS (SELECT 1 FROM "Posts" AS "p" WHERE "b"."BlogId" = "p"."BlogId") AS "Key",
    (SELECT COUNT(*)
        FROM "Blogs" AS "b0"
        WHERE
            EXISTS (SELECT 1 FROM "Posts" AS "p0" WHERE "b"."BlogId" = "p0"."BlogId")
            = EXISTS (SELECT 1 FROM "Posts" AS "p1" WHERE "b0"."BlogId" = "p1"."BlogId")) AS "Count"
FROM "Blogs" AS "b"
GROUP BY
    EXISTS (SELECT 1 FROM "Posts" AS "p" WHERE "b"."BlogId" = "p"."BlogId")
```

The only difference is the `"Count"` calculation, which is now done with an unnecessary subquery. This subquery uses `"b"."BlogId"`, which is not grouped. It could be argued that the whole EXISTS expression **is** actually grouped, but this is at least not supported by PostgreSQL and also leads to an unnecessarily complex SQL.

### Include provider and version information

EF Core version: 6.0.2
Database provider: Microsoft.EntityFrameworkCore.Sqlite, Npgsql.EntityFrameworkCore.PostgreSQL
Target framework: .NET 6.0
Operating system: macOS 12.3
IDE: VS code

**Comment by ajcvickers** [comment]
@sgaertner Presumably you have tested that this is still the behavior with 6.0.6?

**Comment by sgaertner** [comment]
@ajcvickers Yes, the problem still exists with 6.0.6. Using this switch solves the problem:

```C#
AppContext.SetSwitch("Microsoft.EntityFrameworkCore.Issue27102", true);
```

**Comment by roji** [comment]
Pretty sure this is a duplicate of #27266, which was fixed in 7.0.0-preview5. Not sure if anything can be done in 6.0, /cc @smitpatel

**Comment by ajcvickers** [comment]
Looks like #27433 has a workaround.
